src/backend/distributed/executor/multi_utility.c

```diff
@@ -1045,88 +1055,175 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+	HTAB *shardConnectionHash = OpenTransactionsToAllShardPlacements(shardIdList,
+																	relationOwner);
+	ListCell *shardCell = NULL;
+	
+	shardPlacementConnectionList = ConnectionList(shardConnectionHash);
+	MemoryContextSwitchTo(oldContext);
+	
+	if (!isXactCallbackRegistered)
+	{
+		RegisterXactCallback(CompleteShardPlacementTransactions, NULL);
+		isXactCallbackRegistered = true;
+	}
+	
+	foreach(shardCell, shardIdList)
 	{
-		List *shardPlacementList = NIL;
-		ListCell *shardPlacementCell = NULL;
 		uint64 *shardIdPointer = (uint64 *) lfirst(shardCell);
 		uint64 shardId = (*shardIdPointer);
+		ShardConnections *shardConnections = NULL;
+		bool shardConnectionsFound = false;
+		char *escapedCommandString = NULL;
+		StringInfo applyCommand = makeStringInfo();
+
+		shardConnections = GetShardConnections(shardConnectionHash,
+											   shardId,
+											   &shardConnectionsFound);
+		Assert(shardConnectionsFound);
 
 		/* build the shard ddl command */
-		char *escapedCommandString = quote_literal_cstr(commandString);
-		StringInfo applyCommand = makeStringInfo();
+		escapedCommandString = quote_literal_cstr(commandString);
 		appendStringInfo(applyCommand, WORKER_APPLY_SHARD_DDL_COMMAND, shardId,
 						 escapedCommandString);
 
-		shardPlacementList = FinalizedShardPlacementList(shardId);
-		foreach(shardPlacementCell, shardPlacementList)
-		{
-			ShardPlacement *placement = (ShardPlacement *) lfirst(shardPlacementCell);
-			char *workerName = placement->nodeName;
-			uint32 workerPort = placement->nodePort;
+		ExecuteCommandOnShardReplicas(applyCommand, shardId, shardConnections);
 
-			List *queryResultList = ExecuteRemoteQuery(workerName, workerPort,
-													   relationOwner, applyCommand);
-			if (queryResultList == NIL)
-			{
-				/*
-				 * If we failed on the first placement, return false. We return
-				 * here instead of exiting at the end to avoid breaking through
-				 * multiple loops.
-				 */
-				if (isFirstPlacement)
-				{
-					return false;
-				}
+		FreeStringInfo(applyCommand);
+	}
 
-				ereport(WARNING, (errmsg("could not apply command on shard "
-										 UINT64_FORMAT " on node %s:%d", shardId,
-										 workerName, workerPort),
-								  errdetail("Shard placement will be marked as "
-											"inactive.")));
+	/* check for cancellation one last time before returning */
+	CHECK_FOR_INTERRUPTS();
 
-				*failedPlacementList = lappend(*failedPlacementList, placement);
-			}
-			else
-			{
-				ereport(DEBUG2, (errmsg("applied command on shard " UINT64_FORMAT
-										" on node %s:%d", shardId, workerName,
-										workerPort)));
-			}
+	/* CloseConnections(allShardsConnectionList); */
+
+	return true;
+}
+
+
+/*
+ * CompleteShardPlacementTransactions commits or aborts pending shard placement
+ * transactions when the local transaction commits or aborts.
+ */
+static void
+CompleteShardPlacementTransactions(XactEvent event, void *arg)
```

Comment:
I wonder whether it makes sense to move this logic into shard_transaction.c as well, since other callers might want to use it.